File: matrix_class/src/Crout.cpp
#include "../include/LUDecomposition.hpp"
#include <cmath>
using namespace std;

Crout::Crout() : LUDecomposition() {}
Crout::Crout(int r, int c) : LUDecomposition(r, c) {}

SolverResult Crout::solve(double *b, int n, int maxIter, double tol) {
  if (rows != cols || rows != n)
    throw MatrixException("matrix dimensions don't match for Crout LU");

  double **L = new double *[n];
  double **U = new double *[n];
  for (int i = 0; i < n; i++) {
    L[i] = new double[n];
    U[i] = new double[n];
    for (int j = 0; j < n; j++) {
      L[i][j] = 0.0;
      U[i][j] = 0.0;
    }
    U[i][i] = 1.0;
  }

  for (int k = 0; k < n; k++) {
    for (int i = k; i < n; i++) {
      double sum = 0.0;
      for (int s = 0; s < k; s++)
        sum += L[i][s] * U[s][k];
      L[i][k] = data[i][k] - sum;
    }
    for (int j = k + 1; j < n; j++) {
      double sum = 0.0;
      for (int s = 0; s < k; s++)
        sum += L[k][s] * U[s][j];
      if (fabs(L[k][k]) < 1e-12)
        throw MatrixException(
            "zero pivot in Crout... matrix might be singular bro");
      U[k][j] = (data[k][j] - sum) / L[k][k];
    }
  }

  double maxError = 0.0;
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      double sum = 0.0;
      for (int k = 0; k < n; k++)
        sum += L[i][k] * U[k][j];
      double err = fabs(sum - data[i][j]);
      if (err > maxError)
        maxError = err;
    }
  }

  double *y = new double[n];
  for (int i = 0; i < n; i++) {
    double sum = 0.0;
    for (int j = 0; j < i; j++)
      sum += L[i][j] * y[j];
    if (fabs(L[i][i]) < 1e-12)
      throw MatrixException("zero on L diagonal in forward sub... oof");
    y[i] = (b[i] - sum) / L[i][i];
  }

  double *x = new double[n];
  for (int i = n - 1; i >= 0; i--) {
    double sum = 0.0;
    for (int j = i + 1; j < n; j++)
      sum += U[i][j] * x[j];
    x[i] = y[i] - sum;
  }

  for (int i = 0; i < n; i++) {
    delete[] L[i];
    delete[] U[i];
  }
  delete[] L;
  delete[] U;
  delete[] y;

  SolverResult result;
  result.x = x;
  result.n = n;
  result.iterations = 0;
  result.converged = true;
  result.error = maxError;
  return result;
}

File: Extended_in_python/pynumerics/solvers/crout.py
"""
Crout LU Decomposition.

Ported from C++ Crout class. Decomposes A into L and U (unit diagonal).
"""

from pynumerics.matrix import Matrix
from pynumerics.solvers.result import SolverResult
from pynumerics.solvers.lu_base import LUDecomposition
from pynumerics.exceptions import MatrixException


class Crout(LUDecomposition):
    """LU Decomposition using the Crout algorithm (U has unit diagonal)."""

    def solve(self, b: list[float], max_iter: int = 10000, tol: float = 1e-10) -> SolverResult:
        """Solve Ax = b by computing L and U, then substituting.
        
        Args:
            b: Right-hand side vector.
            max_iter: Ignored for direct methods.
            tol: Ignored for direct methods.
            
        Returns:
            SolverResult containing x, L, U, and maximum reconstruction error.
        """
        n = len(b)
        if self.rows != self.cols or self.rows != n:
            raise MatrixException("matrix dimensions don't match for Crout LU")

        L = Matrix(n, n)
        U = Matrix(n, n)
        
        for i in range(n):
            U[i, i] = 1.0

        for k in range(n):
            for i in range(k, n):
                s = sum(L[i, idx] * U[idx, k] for idx in range(k))
                L[i, k] = self._data[i][k] - s
                
            for j in range(k + 1, n):
                s = sum(L[k, idx] * U[idx, j] for idx in range(k))
                if abs(L[k, k]) < 1e-12:
                    raise MatrixException("zero pivot in Crout... matrix might be singular bro")
                U[k, j] = (self._data[k][j] - s) / L[k, k]

        max_error = 0.0
        for i in range(n):
            for j in range(n):
                s = sum(L[i, k] * U[k, j] for k in range(n))
                err = abs(s - self._data[i][j])
                if err > max_error:
                    max_error = err

        y = [0.0] * n
        for i in range(n):
            s = sum(L[i, j] * y[j] for j in range(i))
            if abs(L[i, i]) < 1e-12:
                raise MatrixException("zero on L diagonal in forward sub... oof")
            y[i] = (b[i] - s) / L[i, i]

        x = [0.0] * n
        for i in range(n - 1, -1, -1):
            s = sum(U[i, j] * x[j] for j in range(i + 1, n))
            x[i] = y[i] - s

        return SolverResult(
            x=x,
            n=n,
            iterations=0,
            converged=True,
            dominance_achieved=True,
            error=max_error,
            L=L,
            U=U
        )
